content/docs/user-guide/machine-learning-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,264 @@
+$ dvc exp init './another_script.sh $MYENVVAR'
+```
+
+<admon type="warn">
+
+While DVC is platform-agnostic, commands defined in `dvc.yaml` (`cmd` field) may
+only work on some operating systems and require certain software packages or
+libraries in the environment.
+
+</admon>
+
+Commands are executed sequentially until all are finished or until one of them
+fails (see `dvc repro`).
+
+## Simple dependencies
+
+There's more than one type of stage dependency. A simple dependency is a file or
+directory needed for the stage `cmd` to run successfully. When it's contents
+have changed, DVC "invalidates" the stage -- it knows that it needs to run
+again.
+
+<admon type="info">
+
+DVC [calculates a hash] of file/dir contents to compare vs. previous versions.
+This is a distinctive mechanism over traditional build tools like `make`.
+
+[calculates a hash]:
+  /doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory
+
+</admon>
+
+File system-level dependencies are defined in the `deps` list of `dvc.yaml`;
+Alternatively, using the `--deps` (`-d`) option of `dvc stage add` (see the
+previous section's example).
+
+<details>
+
+### External dependencies: click to learn more.
+
+A less common kind of dependency is a _URL dependency_. Instead of files in a
+local disk, you can `dvc import` data from another <abbr>DVC project</abbr> (for
+example hosted on GitHub). External dependencies establish relationships between
+different projects or systems (see `dvc import-url`).
+[Get all the details](/doc/user-guide/external-dependencies).
+
+<admon type="info">
+
+DVC will use special methods to check whether the contents of an URL have
+changed for the purpose of stage invalidation.
+
+</admon>
+
+</details>
+
+## Parameter dependencies
+
+A more ML-specific type of dependency is the _hyperparameter_ (`params` field).
+These are simple values used inside your code to tune data processing, modeling
+attributes, or that determine stage execution in any other way. For example, a
+[random forest classifier] may require a _maximum depth_ value.
+
+Instead of hard-coding it, your code can read param values from a parameters
+file. `dvc params` can track any key/value pair inside structured YAML, JSON,
+TOML, or Python files (see also `dvc.api.params_show()`). DVC will keep track of
+params as granular dependencies: it will only invalidate the stage if that part
+of the file has changed.
+
+Parameters are defined under the `params` set of `dvc.yaml`. Each entry is the
+key string to look for the param value in the default params file, `params.yaml.
+Sub-list under custom params file names can also be included.
+
+```yaml
+stages:
+  featurize:
+    cmd: ...
+    ...
+    params:
+      - learning_rate  # from params.yaml
+      - deep_learning.json:
+          - epochs     # from custom deep_learning.json file
+```
+
+[random forest classifier]:
+  https://medium.com/all-things-ai/in-depth-parameter-tuning-for-random-forest-d67bb7e920d
```

Comment:
What about outputs? I see a lot of great detail on deps but seems like outputs is missing?

Reply:
> What about outputs?

Leaving it for a separate PR (WIP) which will be in another page of this section of the UG. Originally it was planned to also metrics and plots but not sure about the latter, as it would overlap with your work in https://github.com/iterative/dvc.org/pull/3830...

Reply:
> outputs... will be in another page of this section of the UG. Originally it was planned to also metrics and plots...

Originally this page wasn't called "Defining pipelines" either. So maybe you're right and we can have a single **Outputs** section here that links to separate metrics and plots guides in the future. TBH Idk 😵‍💫

Reply:
So is the plan to have a follow-up PR with the outputs soon? Just want to make sure there's a plan to add that info before merging this one.

Reply:
> if this page has deps it should have outs

Outputs are mentioned in the **Stages** section for now but I'm still not sure whether they'll end up in this same page or in a separate one.

> So is the plan to have a follow-up PR with the outputs soon? 

Yes, the master plan is 3-4 pages: this one (**Defining Pipelines**), **Reproducing Pipes**, **Pipeline Ops** (or "Operationalizing"?), and maybe **Experiment(al) Pipes**. WDYT @dberenbaum ?

Reply:
@dberenbaum I ended up including **Outputs** here after moving the **Stage command** section to the dvc.yaml guide/ref. PTAL.